### ISSUE TITLE
PHOTON small (200M) の本格学習と val loss 確認

### ISSUE THREAD

**Issue by Kewton** [Kewton/photon-mlx#4]
## 概要

PHOTON tiny (79M) の学習は 100 step テストで loss 70.4% 低減を確認済み。次は small (200M) で本格学習を実行し、multi-turn working memory の有効性を確認する。

> **注意**: 以下の実行計画を実行するには、先にコードベースのリファクタリング（TrainingConfig 導入、gradient accumulation 実装）とデータ準備が必要。

---

## 前提

- 学習コーパス: `data/processed/train_small.jsonl`（生成後に統計確認）
- Config: `configs/photon_small.yaml`
  - hidden_size: 1024, intermediate: 2816, heads: 16, enc/dec: 3 layers each
  - context_length: 2048, micro_batch: 2, grad_accum: 16 → effective batch: 32
  - max_steps: 12000, LR: 1.5e-4（warmup_ratio=0.03, min_lr=1.5e-5）
  - weight_decay: 0.1, max_grad_norm: 1.0
  - checkpoint: 1000 step ごと、eval: 250 step ごと

### 実行環境要件
- Apple Silicon（推奨: 36GB 以上の統合メモリ）
- small モデルで context_length=2048 + micro_batch=2 の場合、推定ピークメモリ ~20–30GB（モデル 400MB fp16 + 活性化 + 勾配 + Adam 状態）
- OOM 時フォールバック: batch_size=1 または context_length=1024 に削減

### 既知の設定不一致（対応必要）
- **[Critical] trainer.py のハードコード問題**: 現状 `trainer.py` 内の `lr`, `batch_size`, `max_steps` 等がハードコードされており YAML 設定が無視される。実行前に TrainingConfig 実装が必須。
- **[Critical] gradient accumulation 未実装**: 現状の trainer.py に grad_accum の実装がなく、effective batch size 32 が実現されない。
- **[Critical] 学習データ未生成**: `data/processed/train_small.jsonl` は未生成。データ準備フェーズが必要。

---

### Phase 0: コードリファクタリング（実行前提）

**0-1. TrainingConfig の導入（`torch_ref/config.py` + `photon_mlx/trainer.py`）**
- `TrainingConfig` データクラスを追加（learning_rate, min_learning_rate, warmup_ratio, micro_batch_size, gradient_accumulation_steps, weight_decay, max_grad_norm, max_steps, eval_every_steps, save_every_steps, log_every_steps, train_corpus, val_corpus 等）
- `PhotonConfig` に `training: Optional[TrainingConfig] = None` として追加（後方互換性維持: training セクションが YAML に存在しない場合は None のまま動作し、既存テスト・既存コードへの影響ゼロ）
- `load_photon_config()` で training セクションをパース

**0-2. `trainer.py` のリファクタリング**
- `train()` 関数のシグネチャを `cfg: PhotonConfig` のみに簡素化し、train_corpus/val_corpus/checkpoint_dir/log_dir は `cfg.training` から読む
- `scripts/train_photon.py` も `cfg` を渡すだけのシンプルな呼び出しに変更
- ハードコード値（lr, batch_size, max_steps 等）を `cfg.training.*` に置き換え
- **gradient accumulation 実装**:
  - MLX の `nn.value_and_grad` を引き続き使用し、micro_batch ごとに grad を取得 → 手動加算
  - `gradient_accumulation_steps` 回後に勾配を平均化 → `clip_grad_norm` → `optimizer.update` → `mx.eval` の順で実装
- **LR スケジューラ実装**:
  - `mlx.optimizers.join_schedules(schedules=[linear_schedule(init=0, end=lr, steps=warmup_steps), cosine_decay(init=lr, decay_steps=total_steps-warmup_steps, end=min_lr)], boundaries=[warmup_steps])` を使用
  - `AdamW(learning_rate=schedule_fn)` で構築
- **AdamW + weight_decay + gradient clipping** の追加
- **checkpoint 後方互換性**:
  - `state.json` の追加フィールドは load 時に KeyError にならないよう `TrainState` のデフォルト値で吸収
  - optimizer state は別ファイル（`optimizer.npz`）に保存し、resume 時に存在しなくてもエラーにしない
  - 既存テスト `test_save_and_load`、`test_load_model`、`test_load_model_forward_consistency` が変更なしで通過することを検証
- `train_photon_quick.py` は Phase 0 スコープ外（後続で統合検討）

**0-3. loss breakdown ログの実装**
- `total, _ = photon_loss(...)` の `_` を `breakdown` として取得
- JSONL ログに `{"step": N, "train_loss": X, "next_token_loss": Y}` 形式で記録（v1: next_token のみ）

**Phase 0 完了後に追加する統合テスト（推奨）**:
- tiny config + 小規模ダミーデータで `train()` を 10 step 実行
- (1) loss が減少すること、(2) checkpoint が正しく保存されること、(3) resume が機能すること を検証

---

### Phase 1: データ準備

**1-0. 対象リポジトリ準備**
- `repo.repo_path` は `./data/raw/fastapi` を前提とする
- 未配置なら以下で取得し、学習対象 commit に固定する

```bash
mkdir -p data/raw
git clone https://github.com/fastapi/fastapi.git data/raw/fastapi
git -C data/raw/fastapi checkout eba8942c81dbf990d25fbae34e6601bdbc21e74b
```

**Phase 1 の config 役割分担**
- 現行スクリプトでは `configs/baseline.yaml` が repo/index 定義（`paths.data_root`, `repo.include`, `repo.exclude`, `repo.repo_commit`）を担う
- `configs/photon_small.yaml` は tokenizer / training 意図を担う
- `repo.repo_id`、`repo.repo_commit`、`paths.data_root` は両設定で同期させる

**1-1. リポジトリインジェスト**
```bash
python scripts/ingest_repo.py \
  --repo ./data/raw/fastapi \
  --repo-id fastapi_fastapi \
  --commit eba8942c81dbf990d25fbae34e6601bdbc21e74b \
  --config configs/baseline.yaml
```

**1-2. 学習コーパス生成**
```bash
python scripts/generate_training_corpus.py \
  --repo-id fastapi_fastapi \
  --config configs/baseline.yaml \
  --photon-config configs/photon_small.yaml \
  --output-dir data/processed
```
- 生成されるファイル: `train_small.jsonl`, `val_small.jsonl`（`train_tiny` の先頭 5000 ドキュメントのサブセット）

**1-3. データ量検証**
- 生成後に統計（ドキュメント数、トークン数）を確認
- 必要総トークン数の概算: effective batch 32 × context_length 2048 × 12000 steps ≈ **786M トークン**（エポック回転で賄う前提）
- 不足する場合は max_chunks 制限の緩和またはエポック数の計算を行う

---

### Phase 2: tiny 5000 step 完全版学習

```bash
python scripts/train_photon.py --config configs/photon_tiny.yaml
```
- context_length=2048 での1ステップあたり実測時間を確認してから推定を更新
- val loss 推移を確認

### Phase 3: small 12000 step 本格学習

```bash
python scripts/train_photon.py --config configs/photon_small.yaml
```
- loss breakdown ログ（next_token のみ、v1）を記録
- checkpoint: 1000 step ごとに保存

### Phase 4: 結果分析と記録

- tiny/small の val loss 推移を比較
- loss 崩壊パターンがあれば `failure_cases/` に記録

---

## 成功条件（定量的閾値）

- val loss が最初の eval 時点から最終 eval 時点で **30% 以上** 低下する
- tiny の最終 val loss を small が **下回る**（スケーリング改善）
- loss 崩壊なし（定義: val loss が 1000 step 以上にわたって単調増加するパターンがない）

---

## 推定所要時間

- Phase 0 リファクタリング: 2–4 時間
- Phase 1 データ準備: 30–60 分
- tiny 5000 steps: context_length=2048 での実測後に更新（smoke test は context_length=64 で計測済み）
- small 12000 steps: tiny 実測値から推定

---

## 関連ファイル

- `scripts/train_photon.py` — エントリーポイント（シグネチャ修正対象）
- `photon_mlx/trainer.py` — 訓練ループ（要リファクタリング）
- `torch_ref/config.py` — PhotonConfig（TrainingConfig 追加が必要）
- `photon_mlx/data.py` — データパイプライン
- `photon_mlx/loss.py` — loss 計算（breakdown dict あり）
- `photon_mlx/model.py` — モデル実装
- `configs/photon_tiny.yaml` — tiny 設定
- `configs/photon_small.yaml` — small 設定
- `configs/baseline.yaml` — repo/index 設定（Phase 1 スクリプト入力）
- `scripts/generate_training_corpus.py` — コーパス生成（出力: train_tiny の先頭5000件 → train_small）
- `scripts/generate_photon.py` — 推論スクリプト（load_model に依存、checkpoint 互換性注意）

**Comment by Kewton** [comment]
PR #17 マージ完了。gradient accumulation 実装、small model training config 追加。149テスト全パス。
